src/librustc_target/abi/call/riscv.rs

```diff
@@ -1,49 +1,341 @@
+}
+
+#[derive(Copy, Clone)]
+enum FloatArgPass {
+    FloatPair,
+    Float,
+    MixedPair,
+}
+
+fn is_vector<'a, Ty>(arg: &ArgAbi<'a, Ty>) -> bool {
+    match arg.layout.abi {
+        Abi::Vector { .. } => true,
+        _ => false,
+    }
+}
+
+fn is_aggregate<'a, Ty>(arg: &ArgAbi<'a, Ty>) -> bool {
+    arg.layout.is_aggregate() || is_vector(arg)
+}
+
+fn should_use_fp_conv_helper<'a, Ty, C>(
+    cx: &C,
+    arg_layout: &TyLayout<'a, Ty>,
+    xlen: u64,
+    flen: u64,
+    field1_kind: &mut FieldKind,
+    field2_kind: &mut FieldKind,
+) -> bool
+where
+    Ty: TyLayoutMethods<'a, C> + Copy,
+    C: LayoutOf<Ty = Ty, TyLayout = TyLayout<'a, Ty>>,
+{
+    match arg_layout.abi {
+        Abi::Scalar(ref scalar) => match scalar.value {
+            abi::Int(..) | abi::Pointer => {
+                if arg_layout.size.bits() > xlen {
+                    return false;
+                }
+                match (*field1_kind, *field2_kind) {
+                    (FieldKind::Unknown, _) => {
+                        *field1_kind = FieldKind::Integer;
+                        true
+                    }
+                    (FieldKind::Float, FieldKind::Unknown) => {
+                        *field2_kind = FieldKind::Integer;
+                        true
+                    }
+                    _ => false,
+                }
+            }
+            abi::F32 | abi::F64 => {
+                if arg_layout.size.bits() > flen {
+                    return false;
+                }
+                match (*field1_kind, *field2_kind) {
+                    (FieldKind::Unknown, _) => {
+                        *field1_kind = FieldKind::Float;
+                        true
+                    }
+                    (_, FieldKind::Unknown) => {
+                        *field2_kind = FieldKind::Float;
+                        true
+                    }
+                    _ => false,
+                }
+            }
+        },
+        Abi::Vector { .. } | Abi::Uninhabited => false,
+        Abi::ScalarPair(..) | Abi::Aggregate { .. } => {
+            let is_union = match arg_layout.fields {
+                FieldPlacement::Union(_) => true,
+                FieldPlacement::Array { count, .. } => {
+                    for _ in 0..count {
+                        let elem_layout = arg_layout.field(cx, 0);
+                        if !should_use_fp_conv_helper(
+                            cx,
+                            &elem_layout,
+                            xlen,
+                            flen,
+                            field1_kind,
+                            field2_kind,
+                        ) {
+                            return false;
+                        }
+                    }
+                    return true;
+                }
+                FieldPlacement::Arbitrary { .. } => false,
+            };
+            if is_union && arg_layout.size.bits() > 0 {
+                return false;
+            }
```

Comment:
You can check for a ZST with `arg_layout().is_zst()`, I think?

Also, if only ZST unions are allowed, you don't need to check the fields, right?
No scalar/vector fields could ever be found in a DST.

You could instead have `FieldPlacement::Union(_) => return arg_layout.is_zst(),`, maybe?